vcd/resource_vcd_solution_add_on_instance.go

```diff
@@ -0,0 +1,222 @@
+package vcd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceVcdSolutionAddonInstance() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceVcdSolutionAddonInstanceCreate,
+		ReadContext:   resourceVcdSolutionAddonInstanceRead,
+		UpdateContext: resourceVcdSolutionAddonInstanceUpdate,
+		DeleteContext: resourceVcdSolutionAddonInstanceDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceVcdSolutionAddonInstanceImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"add_on_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Parent Solution Add-On ID",
+			},
+			"accept_eula": {
+				Type:        schema.TypeBool,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Defines if EULA is accepted. `false` will and print EULA",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Solution Add-On Instance Name",
+			},
+			"validate_only_required_inputs": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Defines if all or only required inputs should be validated",
+			},
+			"input": {
+				Type:        schema.TypeMap,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "Key value map of Solution Add-On Instance",
+			},
+			"delete_input": { // These will only be applicable to "delete" operation
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Description: "Key value map for deletion of Solution Add-On Instance",
+			},
+			"rde_state": {
+				Type:        schema.TypeString,
+				Description: "Parent RDE state",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func resourceVcdSolutionAddonInstanceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+	addOn, err := vcdClient.GetSolutionAddonById(d.Get("add_on_id").(string))
+	if err != nil {
+		return diag.Errorf("error retrieving Solution Add-On: %s", err)
+	}
+
+	if addOn.SolutionAddOnEntity.Eula != "" && !d.Get("accept_eula").(bool) {
+		return diag.Errorf("cannot create Solution Add-On Instance without accepting EULA.\n\n%s\n\n: %s", addOn.SolutionAddOnEntity.Eula, err)
+	}
+
+	// making a copy of `input` map because mutating it causes terraform state errors
+	input := d.Get("input")
+	inputMap := input.(map[string]interface{})
+	inputCopy := make(map[string]interface{})
+
+	for k, v := range inputMap {
```

Comment:
Since we require go 1.22, can we use `maps.Clone` for this operation?

Reply:
I was going to, but I also have to to prefix `input-%s` at the same time so I would still have to loop.